### ISSUE TITLE
Agent pool maintenance job running multiple times

### ISSUE THREAD

**Issue by AntonyG22** [microsoft/azure-pipelines-agent#2386]
I had my Agent pool maintenance job scheduled for 00:00 UTC, and it was running fine.

For some reason, I wanted the agent job to run now, so I updated the schedule e.g. 15:00, let it run, and then set it back to 00:00.

Now the Maintenance history shows it running 3 times per day across all agents in the pool.
00:00 (twice)
15:00

So I disabled the maintenance job, and re-enabled it. Now the Maintenance history shows it running 4 times per day across all agents in the pool.
00:00 (3 times)
15:00

I'm at a loss how to "reset it" and only run once per day at the scheduled time.

Am I doing something wrong? Can anyone help?

<img width="1585" alt="Screenshot 2019-08-13 at 09 22 49" src="https://user-images.githubusercontent.com/54059217/62926224-f93b9680-bdab-11e9-859a-f7fb95f6c1db.png">

Thanks

Antony

**Comment by juliobbv** [comment]
Hi @AntonyG22,

Are you using an on-prem TFS/Azure DevOps Server? If so, can you run this SQL query for me on SSMS against the configuration database (TFS: `TFS_Configuration`, Azure DevOps Server: `AzureDevOps_Configuration`), and provide the contents of the result query back?

```sql
SELECT  JobId, Data, EnabledState, Flags, LastExecutionTime
FROM    tbl_JobDefinition
WHERE   PartitionId = 1
        AND JobName = 'Pool maintenance schedule job'
```

If you're using hosted Azure DevOps, please provide the name of your account (organization), so we can run the query on our side.

Thanks,
Julio

**Comment by AntonyG22** [comment]
Hi Julio

We're using Azure DevOps. Organisation=bttv

Thanks

**Comment by juliobbv** [comment]
Thanks Antony, I'll take a look.

**Comment by juliobbv** [comment]
Thanks for reporting this issue, @AntonyG22!

We've found a bug in the agent pool maintenance job UI, which was causing extra job schedule records to be created when the intent is to update the existing record. We're filed a bug on the Azure Pipelines team, and we'll get a fix ready to avoid future re-occurrences.

Now, the fix above won't take care of the existing duplicated schedules, but there's a workaround to get you to a good state -- you'll need a REST client of your choice as a pre-requisite, like [Restlet Client](https://restlet.com/modules/client/):

1. From the REST client, issue a GET request to `http://<orgName>.visualstudio.com/_apis/distributedtask/pools/<poolId>/maintenancedefinitions/?api-version=5.2-preview`. Replace `<orgName>` with your organization name, and `<poolId>` with the id of the pool, which can be found in the URL query of the Maintenance job UI page
2. This will give you back a JSON array of maintenance definitions targeting the affected pool. Drill into each definition, and keep track of the id values:
![image](https://user-images.githubusercontent.com/2474119/63045035-80b50280-be9d-11e9-8e70-1e569d300904.png)
3. For each maintenance definition, issue a DELETE request to `http://<orgName>.visualstudio.com/_apis/distributedtask/pools/<poolId>/maintenancedefinitions/<definitionId>?api-version=5.2-preview`. Replace `<definitionId>` with the ids you found in step 2.

After you've gone over all the steps, you should have no scheduled definitions. Then, you can create a new one and you should be all set!

Let me know if you have any questions.

Thanks,
Julio

**Comment by AntonyG22** [comment]
Hi Julio

Thanks for getting back to me.

Struggling to implement your workaround though.

From step 1, I get a response back very similar to your example, so it only includes a single item in the values array, with id: 1

I still tried the DELETE request, adding the 1 in place of the definitionId, and I get a 200 response, but nothing appears to have been deleted.

Repeating step 1 still shows the same single item id=1 in the values array. So it doesn't appear that the DELETE rest API orks for me, but then also doesn't appear that I have multiple maintenance definitions anyway.

Am I missing something?

Here is my response from Step 1.

`{
"count": 1,
"value": [
{
"options": {
"workingDirectoryExpirationInDays": 35
},
"retentionPolicy": {
"numberOfHistoryRecordsToKeep": 10
},
"scheduleSetting": {
"scheduleJobId": "ec1e08be-1478-4c89-94aa-a659a1749a77",
"timeZoneId": "UTC",
"startHours": 0,
"startMinutes": 0,
"daysToBuild": "all"
},
"id": 1,
"pool": {
"id": 15,
"isHosted": false,
"poolType": "automation",
"size": 0,
"isLegacy": null
},
"enabled": true,
"jobTimeoutInMinutes": 60,
"maxConcurrentAgentsPercentage": 25
}
]
}`

**Comment by juliobbv** [comment]
@AntonyG22, I'd double check the URL query for step 2. You should be getting a 204 No Content response from the DELETE request. How do your URLs look like for both steps?

BTW, It's OK if there's only one definition. It's just that that particular definition contains multiple job schedules.

**Comment by AntonyG22** [comment]
Thank you.

My issue was that I couldn't get the REST client authenticating properly for the DELETE request, so tried it in a little PHP script where I was already calling other VS APIs.

I scrapped that idea and persisted with the REST Client and got the DELETE request working and it is as you described.

I've now created a new maintenance definition and  it has a new ID, so fingers crossed that has got me back to a sensible position. I'll find out overnight tonight.

Thanks very much for your help.

Regards

Antony
